code/datums/components/singularity.dm

```diff
@@ -203,13 +209,32 @@
 
 	turfs_to_eat--
 
-/datum/component/singularity/proc/move()
-	var/drifting_dir = pick(GLOB.alldirs - last_failed_movement)
-
-	if(!QDELETED(target) && prob(CHANCE_TO_MOVE_TO_TARGET))
-		drifting_dir = get_dir(parent, target)
-
-	step(parent, drifting_dir)
+/datum/component/singularity/proc/move(delta_time)
+	var/list/drifting_dir = GLOB.alldirs
+	drifting_dir -= last_failed_movement
+	var/singularity_sight_modifier
+	var/nearest_distance = SINGULARITY_SIGHT_SIZE + singularity_sight_modifier
+	var/closest_target
+	var/static/things_to_eat = typecacheof(list(/turf/closed/wall, /obj/structure, /mob/living))
+	var/static/things_to_not_eat = typecacheof(list(/obj/structure/grille/indestructable, /obj/structure/window/reinforced/fulltile/indestructable, /turf/closed/indestructible, /turf/open/indestructible))
+	for(var/atom/A as() in oview(SINGULARITY_SIGHT_SIZE + singularity_sight_modifier, parent)) //Find the nearest wall, strucutre or mob
+		if((A.type in things_to_eat) && !(A.type in things_to_not_eat)) //can we eat it?
+			var/dist = get_dist(parent, A) //Get the distance to it.
+			if(dist < nearest_distance) //Is it closer than the previous iteration?
+				closest_target = A
+				nearest_distance = dist
+				singularity_sight_modifier = 0
+	var/dir = get_dir(parent, closest_target) //go that way
+	if ((dir == null) && (singularity_sight_modifier < 45)) //Increase the search radius if you find nothing, up to a maximum, ofcourse.
+		singularity_sight_modifier++
+		return
+	if(dir in drifting_dir) //Can we go that direction?
+		final_dir = dir //If so, go that way.
+	else
+		final_dir = pick(drifting_dir) //Else, pick a random other direction to go.
+	if(!QDELETED(target) && prob(CHANCE_TO_MOVE_TO_TARGET)) //Do we have a beacon to go to? If so, with chance, go there instead.
+		final_dir = get_dir(parent, target)
+	step(parent, final_dir)
```

Comment:
This is ridiculously expensive. I would recommend removing this and allowing the singularity to freely move, as the singularity beacon exists (plus it can't leave the station for too long anyway since it will always loop back around). The singularity is dangerous enough as it is